### PR TITLE
Spectral norm improvements

### PR DESCRIPTION
**Pull request by t-vi**
- Don't do iterations on weight in eval mode
  To facilitate this, register weight as buffer in order to be able
  to use module with spectral norm in eval mode after immediately
  after loading state dict (#8208)
- Use weight instead of weight_orig as weight when removing
  spectral norm
- Add dim parameter in case the normalization should occur w.r.t.
  a dimension other than 0 (#7865)

@SsnL is this about what you had in mind?

I have to add tests. What would be resonable here
- test that three calls in eval mode give the exact same thing and after remove
- test convergence of SN in the right dimension using fixed random weights for Linear and ConvTranspose2d?

### CODE REVIEW

File: torch/nn/utils/spectral_norm.py [hunk 1]
```diff
@@ -34,40 +40,41 @@ def compute_weight(self, module):
         return weight, u
 
     def remove(self, module):
-        weight = module._parameters[self.name + '_orig']
+        weight = getattr(module, self.name)
         delattr(module, self.name)
         delattr(module, self.name + '_u')
         delattr(module, self.name + '_orig')
-        module.register_parameter(self.name, weight)
+        module.register_parameter(self.name, torch.nn.Parameter(weight))
```
Comment (crcrpar) [inline torch/nn/utils/spectral_norm.py]: Isn't `weight` Parameter? (plz ignore this freely)

**Comment by SsnL** [comment]
Code looks reasonable to me. Thanks! 

I think for the tests you proposed, 1 looks good. It might be hard to do a convergence test for 2. How about just test that `_u` has correct length? :)

File: torch/nn/utils/spectral_norm.py (hunk 1 above)
Comment (t-vi) [reply torch/nn/utils/spectral_norm.py]: While you have spectral norm (which you still do at this point), it is a Tensor (because it is computed / a buffer).

**Comment by t-vi** [comment]
Thanks! Here are the tests.

**Comment by t-vi** [comment]
@ezyang The failures are a timeout and a strange jit fault in LSTM (which I didn't get near to, I hope), also the caffee rocm seems to have completed, but it is not recognized as completed (the latter seems common today)... Do you have any advice for me?

File: test/test_nn.py [hunk 2]
```diff
@@ -1452,6 +1452,26 @@ def test_spectral_norm(self):
         self.assertTrue(hasattr(m, 'weight'))
         self.assertTrue('weight' in m._parameters)
 
+    def test_spectral_norm_eval(self):
```
Comment (SsnL) [inline test/test_nn.py]: Could you also test somewhere that 
1. output doesn't change after adding and immediately removing SN
2. after adding SN, fwd a couple iteration, output doesn't change after removing SN.

**Comment by t-vi** [comment]
"Aborted by user anonymous"?

**Comment by SsnL** [comment]
@t-vi lol CI probably had some problem then. I restarted CI on this. Sorry!

**Review by ssnl — APPROVED** [review]
Looks great. Thanks @t-vi !

**Review by soumith — APPROVED** [review]
